README.md
# Radware CWAF Enrichment - SplunkApp

This project contains a Splunk App that integrates with the Radware Cloud WAF Rest API.

## Getting Started

### App Requirements
* Splunk 9.0.0 or SplunkCloud
* Rest API Credentials for Radware Cloud WAF

### Development Requirements
* NodeJS 16.x or higher
* Powershell Core for Build Scripts (With [Invoke-Build](https://github.com/nightroman/Invoke-Build) Module)
* Yarn 1.22.10 or higher
* Local Splunk Enterprise installation (Dockerized development environment is not yet supported)

### Package Layout
* `frontend` - Contains the frontend code for the radware_cwaf_enrichment Splunk App. Written in [React](https://reactjs.org/) with the [Splunk UI Framework](https://splunkui.splunk.com/)
* `splunkapp` - Contains the Splunk App code for the radware_cwaf_enrichment Splunk App
* `build` - Contains the build scripts for the radware_cwaf_enrichment Splunk App

splunkapp/CHANGELOG.md
# Changelog

All notable changes to this app will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.2] - 2023-03-02
### Fixes
- Removed debug logging of sensitive information

## [1.1.1] - 2023-02-21
### Fixes
- Fixed app manifest versioning
- Specify targetWorkloads and supportedDeployments
- Moved splunklib files under /bin directory
- Added Splunk AppInspect to CI/CD build scripts


## [1.1.0] - 2023-02-21
### New Version
- First release
- Includes generating commands, SplunkWeb UI

test/SplunkApp.Base.tests.ps1
Param (
    $SplunkAPIHost = "https://localhost:8089",
    $SplunkUser = "admin",
    $SplunkClearPassword = "newPassword",
    $MaxWaitSeconds = 60,
    $SplunkNormalUser = "raduser1",
    $SplunkLocalPath = "splunkapp/local"
)

Describe "SplunkApp" {
    BeforeAll {
        $EnvironmentParamArray = @( "SplunkAPIHost", "SplunkURL", "SplunkUser", "SplunkClearPassword", "SplunkNormalUser", "RunningHeadless")
        foreach($EnvironmentParam in $EnvironmentParamArray) {
            if(Test-Path env:$EnvironmentParam) {
                Write-Host "Setting $EnvironmentParam from environment variable."
                Set-Variable -Scope Script -Name $EnvironmentParam -Value (Get-Item env:$EnvironmentParam).Value
                Set-Variable -Scope Global -Name $EnvironmentParam -Value (Get-Item env:$EnvironmentParam).Value
                Set-Variable -Scope Local -Name $EnvironmentParam -Value (Get-Item env:$EnvironmentParam).Value
            }
        }
        Write-Host "Testing SplunkApp on $SplunkAPIHost"
        try{
            if (Test-Path $SplunkLocalPath)
            {
                Remove-Item "$SplunkLocalPath/*.conf" -Force
            }
            if (Test-Path "splunkapp/metadata/local.meta")
            {
                Remove-Item "splunkapp/metadata/local.meta" -Force
            }
        }catch{
            Write-Host "Warning cleaning up test files: $($_.Exception.Message)"
        }
        

        [securestring]$SecureSplunkPassword = ConvertTo-SecureString $SplunkClearPassword -AsPlainText -Force
        [pscredential]$Script:SplunkCreds = New-Object System.Management.Automation.PSCredential ($SplunkUser, $SecureSplunkPassword)
        [pscredential]$Script:SplunkRegularCreds = New-Object System.Management.Automation.PSCredential ($SplunkNormalUser, $SecureSplunkPassword)

        $BaseRequest = @{
            Method = "POST"
            Uri = "$( $SplunkAPIHost )/servicesNS/nobody/radware_cwaf_enrichment/radware_cwaf_enrichment/radware_cwaf_enrichment_config/settings"
            SkipCertificateCheck = $true
            Credential = $SplunkCreds
        }

        function Refresh-SplunkApp
        {

            $refreshUri = "$SplunkAPIHost/services/apps/local/radware_cwaf_enrichment?refresh=true"
            Invoke-RestMethod -Uri $refreshUri -Method GET -Credential $Script:SplunkCreds -SkipCertificateCheck
        }

        function Invoke-SplunkSearchCommand
        {
            Param(
                $SearchQuery = '',
                [pscredential]$Credentials = $SplunkRegularCreds,
                $SplunkAPIHost = $SplunkAPIHost,
                $MaxWaitSeconds = $MaxWaitSeconds
            )
            $SearchResults = 0
            $LoopCounter = 0
            $SearchParams = @{
                search = $SearchQuery
                output_mode = 'json'
                earliest_time = '-5m'
                exec_mode = 'oneshot'
            }
            While (($SearchResults -eq 0) -and ($LoopCounter -lt $MaxWaitSeconds))
            {
                $RequestSplat = @{
                    Method = "POST"
                    Body = $SearchParams
                    Uri = "$( $SplunkAPIHost )/services/search/jobs"
                    SkipCertificateCheck = $true
                    Credential = $Credentials
                }
                $SplunkSearchResults = Invoke-RestMethod @RequestSplat
                $LoopCounter += 1
                $SearchResults = $SplunkSearchResults.results.Count
                Start-Sleep -Seconds 1
            }
            Write-Verbose "Splunk Search Completed - $( $SearchResults ) Events Found"
            return $SplunkSearchResults
        }

        function Ensure-NormalUser
        {

            $RequestSplat = @{
                Method = "GET"
                Uri = "$( $SplunkAPIHost )/services/authentication/users/$( $SplunkNormalUser )"
                SkipCertificateCheck = $true
                Credential = $SplunkCreds
            }
            try
            {
                $UserExists = Invoke-RestMethod @RequestSplat
            }
            catch
            {
                $UserExists = $false
            }
            if (-not$UserExists)
            {
                $RequestSplat = @{
                    Method = "POST"
                    Body = "name=$( $SplunkNormalUser )&roles=user&roles=radware_cwaf_enrichment_admin&password=$( $SplunkClearPassword )&output_mode=json"
                    Uri = "$( $SplunkAPIHost )/services/authentication/users"
                    SkipCertificateCheck = $true
                    Credential = $SplunkCreds
                }
                Invoke-RestMethod @RequestSplat
            }
            else
            {
                $RequestSplat = @{
                    Method = "POST"
                    Body = "&roles=user&roles=radware_cwaf_enrichment_admin"
                    Uri = "$( $SplunkAPIHost )/services/authentication/users/$( $SplunkNormalUser )"
                    SkipCertificateCheck = $true
                    Credential = $SplunkCreds
                }
                Invoke-RestMethod @RequestSplat
            }
        }
    }

    Describe 'custom config endpoint' {
        BeforeAll {
            Ensure-NormalUser
            Refresh-SplunkApp
        }

        Describe 'setting log_level' {
            It 'can update log_level to INFO' {
                $BaseRequest.Body = @{
                    output_mode = "json"
                    "log_level" = "DEBUG"
                }
                $result = Invoke-RestMethod @BaseRequest
                $result.entry[0].content.log_level | Should -Be "DEBUG"
            }

            It 'setting the log_level to an invalid option should fail' {
                $BaseRequest.Body = @{
                    output_mode = "json"
                    "log_level" = "XXCV"
                }
                try
                {
                    Invoke-RestMethod @BaseRequest
                }
                catch
                {
                    $_.Exception.Response.StatusCode | Should -Be 400
                    $_.ErrorDetails.Message | Should -BeLike "*is not in permitted options*"
                }
            }
        }


        Describe 'configuring credential configurations' {
            It 'should raise an error when mandatory parameter is not provided for a new user' {
                $BaseRequest.Body = @{
                    output_mode = "json"
                    "credential.create.username" = "user123"
                    "credential.create.action" = "create"
                }
                try
                {
                    Invoke-RestMethod @BaseRequest
                }
                catch
                {
                    $_.Exception.Response.StatusCode | Should -Be 400
                    $_.ErrorDetails.Message | Should -BeLike "*Missing required parameter*"
                }
            }

            It 'should successfully create a new user' {
                $BaseRequest.Body = @{
                    output_mode = "json"
                    "credential.create.username" = "mock_user_radware_prod"
                    "credential.create.name" = "mock_radware_prod"
                    "credential.create.password" = "xxx"
                    "credential.create.action" = "create"
                }
                $result = Invoke-RestMethod @BaseRequest
                $result.entry[0].content | Should -Not -BeNullOrEmpty
            }

            It 'should successfully update an existing user' {
                $Uname = (New-TimeSpan -Start (Get-Date "01/01/1970") -End (Get-Date)).TotalSeconds
                $BaseRequest.Body = @{
                    output_mode = "json"
                    "credential.1.username" = "mock_$Uname"
                    "credential.1.password" = "mock_$Uname"
                    "credential.1.action" = "update"
                }
                $result = Invoke-RestMethod @BaseRequest
                $result.entry[0].content.'credential.1.username' | Should -Be "mock_$Uname"
            }

            It 'should not update existing user properties if action is not set to update' {
                $Uname = (New-TimeSpan -Start (Get-Date "01/01/1970") -End (Get-Date)).TotalSeconds
                $BaseRequest.Body = @{
                    output_mode = "json"
                    "credential.1.username" = $Uname
                }
                $result = Invoke-RestMethod @BaseRequest
                $result.entry[0].content.'credential.1.username' | Should -Not -Be $Uname
            }

            It 'should successfully delete a user and their passwords' {
                $BaseRequest.Body = @{
                    output_mode = "json"
                    "credential.1.action" = "delete"
                }
                $result = Invoke-RestMethod @BaseRequest
                $result.entry[0].content.'credential.1.password' | Should -BeNullOrEmpty
            }
        }
    }
    Describe 'Generating Commands' {
        BeforeAll {
            Ensure-NormalUser
            Refresh-SplunkApp
            $BaseRequest.Body = @{
                output_mode = "json"
                "credential.create.username" = "mock_user_radware_prod"
                "credential.create.name" = "mock_radware_prod"
                "credential.create.password" = "fake_password"
                "credential.create.action" = "create"
            }
            Invoke-RestMethod @BaseRequest | Out-Null
        }

        It 'should execute radwarecwaflistremote command ' {
            $result = Invoke-SplunkSearchCommand -SearchQuery " | radwarecwaflistremote"
            $result.results.Count | Should -BeExactly 3
        }

        It 'should execute radwarecwafimportremote command ' {
            $result = Invoke-SplunkSearchCommand -SearchQuery " | radwarecwafimportremote"
            $result.results.Count | Should -BeExactly 4
        }

        It 'should execute radwarecwafdeletelocal command ' {
            $result = Invoke-SplunkSearchCommand -SearchQuery " | radwarecwafdeletelocal"
            $result.results.Count | Should -BeExactly 1
        }

        AfterAll {
            $BaseRequest.Body = @{
                output_mode = "json"
                "credential.1.action" = "delete"
            }
            Invoke-RestMethod @BaseRequest | Out-Null
        }
    }
}



test/frontend/SplunkApp.Selenium.tests.ps1
Param (
    $SplunkAPIHost = "https://localhost:8089",
    $SplunkURL = "http://localhost:8000",
    $SplunkUser = "admin",
    $SplunkClearPassword = "newPassword",
    $MaxWaitSeconds = 60,
    $SplunkNormalUser = "raduser1",
    $RunningHeadless = $false,
    $DemoTest = $true
)
Describe "SplunkApp radware_cwaf_enrichment Frontend" {
    BeforeAll {
        $EnvironmentParamArray = @( "SplunkAPIHost", "SplunkURL", "SplunkUser", "SplunkClearPassword", "SplunkNormalUser", "RunningHeadless")
        foreach ($EnvironmentParam in $EnvironmentParamArray)
        {
            if (Test-Path env:$EnvironmentParam)
            {
                Write-Host "Setting $EnvironmentParam from environment variable."
                Set-Variable -Scope Script -Name $EnvironmentParam -Value (Get-Item env:$EnvironmentParam).Value
                Set-Variable -Scope Global -Name $EnvironmentParam -Value (Get-Item env:$EnvironmentParam).Value
                Set-Variable -Scope Local -Name $EnvironmentParam -Value (Get-Item env:$EnvironmentParam).Value
            }
        }
        function Load-LocalSelenium
        {
            $Script:SeleniumPath = Resolve-Path -Path "$PSScriptRoot\..\..\build\selenium"
            Write-Host "$( $Script:SeleniumPath )\WebDriver.dll"
            [System.Reflection.Assembly]::LoadFrom("$( $Script:SeleniumPath )\WebDriver.dll")
        }

        function Check-ElementExists
        {
            Param(
                [parameter(Position = 0, Mandatory = $true)]$XPathOrId
            )
            try
            {
                return $WebDriver.FindElement([OpenQA.Selenium.By]::Xpath($XPathOrId));
            }
            catch [OpenQA.Selenium.NoSuchElementException]
            {
                try
                {
                    return $WebDriver.FindElement([OpenQA.Selenium.By]::Id($XPathOrId))
                }
                catch [OpenQA.Selenium.NoSuchElementException]
                {
                    return $false
                }
            }
        }

        function Wait-Element
        {
            Param(
                [parameter(Position = 0, Mandatory = $true)]$XPathOrId,
                [parameter(Position = 1, Mandatory = $false)]$Timeout = 10,
                [Switch]$Click = $false,
                [Switch]$WaitForDisappear
            )
            $StopLoop = $false
            $RetryEnd = $Timeout * 1000 / 250
            do
            {
                $element = (Check-ElementExists -XPathOrId $XPathOrId)
                if ($element)
                {
                    if($WaitForDisappear){
                        $StopLoop = $false
                    }else{
                        $StopLoop = $true
                        if ($Click)
                        {
                            $element.Click()
                            return
                        }
                        return $element
                    }
                }else{
                    if($WaitForDisappear){
                        $StopLoop = $true
                    }
                }
                if ($RetryCount -gt $RetryEnd)
                {
                    Write-Error "Waiting for $XPathOrId FAILED"
                    $StopLoop = $true

                }
                $RetryCount++
                Start-Sleep -Milliseconds 250
            } While ($StopLoop -eq $false )
        }

        function Set-InputContent
        {
            Param(
                [parameter(Position = 0)]$XPathOrId,
                [parameter(Position = 1)]$Text
            )
            $el = Check-ElementExists -XPathOrId $XPathOrId
            $el.Clear()
            $el.SendKeys($Text)
        }

        function Start-WebDriver
        {
            $Options = New-Object OpenQA.Selenium.Chrome.ChromeOptions
            if($RunningHeadless){
                $Options.AddArgument("--headless")
            }
            $Options.AddArgument("--log-level=3")
            $Options.AddArguments("--disable-gpu", "--window-size=1920,1200", "--ignore-certificate-errors", "--disable-extensions", "--no-sandbox", "--disable-dev-shm-usage") | Out-Null
            if ($WebDriver)
            {
                if (-not$WebDriver.CurrentWindowHandle)
                {
                    $global:WebDriver = new-object OpenQA.Selenium.Chrome.ChromeDriver($Options)
                }
            }
            else
            {
                $global:WebDriver = new-object OpenQA.Selenium.Chrome.ChromeDriver($Options)
            }
        }

        function Close-WebDriver
        {
            if ($WebDriver)
            {
                Write-Host "Closing Driver"
                if ($WebDriver.CurrentWindowHandle)
                {
                    $WebDriver.Close()
                }
            }
        }

        function Login-SeleniumSplunk
        {
            Write-Host "Logging in to Splunk @ $SplunkURL"
            $WebDriver.Navigate().GoToUrl($SplunkURL)
            if (Check-ElementExists -XPathOrId "//input[@id='username']")
            {
                Set-InputContent -XPathOrId "username" -Text $SplunkUser
                Set-InputContent -XPathOrId "password" -Text $SplunkClearPassword
                $WebDriver.FindElement([OpenQA.Selenium.By]::XPath("//input[@type='submit' and @value='Sign In']")).Click()
            }
            elseif ($WebDriver.Url -like "*/en-US/app/launcher/home")
            {
                Write-Host "Already Logged In"
            }
        }
    }

    AfterAll {
        if($WebDriver) {
            $WebDriver.Dispose()
        }
    }

    if (-not$DemoTest)
    {
        Describe 'basic Splunk acessibility' {
            BeforeAll {
                Load-LocalSelenium
                Start-WebDriver
            }

            It 'logs into Splunk successfully and the app is visible' -Tag 'RunsInSplunk' {
                Login-SeleniumSplunk
                Wait-Element -XPathOrId "//a[@aria-label='Radware CWAF Data Import for Splunk']"
                (Check-ElementExists -XPathOrId "//a[@aria-label='Radware CWAF Data Import for Splunk']") | Should -Not -BeNullOrEmpty
            }

            AfterAll {
                Close-WebDriver
            }
        }
    }

    Describe 'radware_cwaf_enrichment react app test' {
        BeforeAll {
            Load-LocalSelenium
            Start-WebDriver
            if (-not$DemoTest)
            {
                Login-SeleniumSplunk
                Wait-Element -XPathOrId "//a[@aria-label='Radware CWAF Data Import for Splunk']"
            }
        }

        It 'this will fail' {
            $True | Should -Be $False
        }

        It 'successfully loads the start page' {
            if (-Not$DemoTest)
            {
                $WebDriver.Navigate().GoToUrl("$SplunkURL/en-US/app/radware_cwaf_enrichment/start")
            }
            else
            {
                $WebDriver.Navigate().GoToUrl("$SplunkURL")
            }
            Wait-Element -XPathOrId "//button[@title='Setup']"
            $SetupButton = Check-ElementExists -XPathOrId "//button[@title='Setup']"
            $SetupButton | Should -Not -BeNullOrEmpty
        }

        It 'successfully loads the setup tab' {
            $SetupButton = Check-ElementExists -XPathOrId "//button[@title='Setup']"
            $SetupButton | Should -Not -BeNullOrEmpty
            $SetupButton.Click()
        }

        It 'successfully loads the manage objects tab' {
            $TabButton = Check-ElementExists -XPathOrId "//button[@title='Manage Objects']"
            $TabButton | Should -Not -BeNullOrEmpty
            $TabButton.Click()
        }

        It 'successfully loads the about tab' {
            $TabButton = Check-ElementExists -XPathOrId "//button[@title='About']"
            $TabButton | Should -Not -BeNullOrEmpty
            $TabButton.Click()
        }

        It 'successfully loads the logs tab' {
            $TabButton = Check-ElementExists -XPathOrId "//button[@title='Logs']"
            $TabButton | Should -Not -BeNullOrEmpty
            $TabButton.Click()
        }

        Describe 'settings page' {
            BeforeAll {
                $Script:CredentialIndex = $( Get-Random -Minimum 100 -Maximum 999 )
                $Script:CredentialName = "SelTest$( $CredentialIndex )"
                $Script:CredentialUsername = "mock_sel$( $Script:CredentialIndex )"
            }
            It 'successfully changes the log level' {
                Wait-Element -XPathOrId "//button[@title='Setup']" -Click
                $SelectButton = Wait-Element -XPathOrId "//button[@data-test-id='log-level-select']"
                $SelectButton.Click()
                $SelectOption = Wait-Element -XPathOrId "//button[@data-test-id='log-level-select-option-ERROR']"
                $SelectOption.Click()
                $SaveButton = Wait-Element -XPathOrId "//button[@data-test-id='save-settings-button']"
                $SaveButton.Click()
            }

            It 'adds a new credential' {
                Wait-Element -XPathOrId "//button[@title='Setup']" -Click
                Wait-Element -XPathOrId "//button[@data-test-id='add-new-credential-button']" -Click
                Wait-Element -XPathOrId "//div[@data-test-id='credential-credential_name']/span/input"
                Set-InputContent -XPathOrId "//div[@data-test-id='credential-credential_name']/span/input" `
                    -Text $CredentialName
                Set-InputContent -XPathOrId "//div[@data-test-id='credential-username']/span/input" `
                    -Text $CredentialUsername
                Set-InputContent -XPathOrId "//div[@data-test-id='credential-password']/span/input" `
                    -Text "password"
                Set-InputContent -XPathOrId "//div[@data-test-id='credential-password_confirm']/span/input" `
                    -Text "password"
                Wait-Element -XPathOrId "//button[@data-test-id='credential-save-button']" -Click
                Wait-Element -XPathOrId "//td[@data-test='cell' and text()='$($CredentialName)']"
            }

            It 'Deletes a credential' {
                Wait-Element -XPathOrId "//td[@data-test='cell' and text()='$($CredentialName)']"
                Wait-Element -XPathOrId "//button[@data-test-id = 'delete-$($CredentialName)-button']" -Click
                Wait-Element -XPathOrId "//button[@data-test-id = 'confirm-delete-credential-$($CredentialName)-button']" -Click
                Wait-Element -XPathOrId "//td[@data-test='cell' and text()='$($CredentialName)']" -WaitForDisappear
            }
        }

        # AfterAll {
        #     Close-WebDriver
        # }
    }
}

InvokeBuild.build.ps1
param(
    $SplunkAPIHost = "https://localhost:8089",
    $SplunkURL = "http://localhost:8000",
    $SplunkUser = "admin",
    $SplunkNormalUser = "raduser1",
    $SplunkAppName = "radware_cwaf_enrichment",
    $SplunkContainerName = "splunkdev-web-1",
    $SplunkClearPassword = "newPassword",
    $MaxWaitSeconds = 240,
    $BuildRoot = $BuildRoot,
    ## Docker Variables
    $DockerSplunkImage = "splunk/splunk:9.0.3",
    $DockerSplunkHostname = "splunkdev001",
    $DockerSplunkWebPort = "9000",
    $DockerSplunkAPIPort = "19089",
    $DockerSplunkHECPort = "19088",
    $DockerSplunkAppSrc = "../splunkapp",
    ## Splunkbase Variables
    $MaxWaitAppInspect = 300,
    [switch]$TestFrontendInDemo = $false
)

task SetupVariables {
    $Script:SplunkAppPath = (Resolve-Path "$( $BuildRoot )/splunkapp/")
    $Script:FrontendPath = (Resolve-Path "$( $BuildRoot )/frontend/")
    [securestring]$SecureSplunkPassword = ConvertTo-SecureString $SplunkClearPassword -AsPlainText -Force
    [pscredential]$Script:SplunkCreds = New-Object System.Management.Automation.PSCredential ($SplunkUser, $SecureSplunkPassword)
    $Script:RequestSplat = @{
        Method               = "GET"
        Uri                  = "$( $SplunkAPIHost )/services/authentication/users/$( $SplunkUser )"
        SkipCertificateCheck = $true
        Credential           = $SplunkCreds
    }

    $Script:PesterConfig = New-PesterConfiguration
    $Script:PesterConfig.TestResult.OutputFormat = "NUnitXml"
    $Script:PesterConfig.TestResult.OutputPath = "$( $BuildRoot )/output/TestResults.xml"
    $Script:PesterConfig.TestResult.Enabled = $True
    $Script:PesterConfig.Output.Verbosity = "Detailed"
    $Script:PesterConfig.Run.PassThru = $True

    $EnvironmentParamArray = @( "SplunkAPIHost", "SplunkURL", "SplunkUser", "SplunkClearPassword", "SplunkNormalUser")
    $EnvironmentParamArray += @( "DockerSplunkImage", "DockerSplunkHostname", "DockerSplunkWebPort", "DockerSplunkAPIPort", "DockerSplunkHECPort", "SplunkContainerName", "DockerSplunkAppSrc")
    foreach ($EnvironmentParam in $EnvironmentParamArray) {
        if (Test-Path env:$EnvironmentParam) {
            Write-Host "Setting $EnvironmentParam from environment variable."
            Set-Variable -Scope Script -Name $EnvironmentParam -Value (Get-Item env:$EnvironmentParam).Value
            Set-Variable -Scope Local -Name $EnvironmentParam -Value (Get-Item env:$EnvironmentParam).Value
            Set-Variable -Scope Global -Name $EnvironmentParam -Value (Get-Item env:$EnvironmentParam).Value
        }
    }
}

task UpdateApp SetupVariables, {
    $destinationPath = "$( $env:SPLUNK_HOME )/etc/apps/$( $SplunkAppName )"
    if (-Not(Test-Path $destinationPath)) {
        New-Item -Path $destinationPath -ItemType SymbolicLink -Value $SplunkAppPath
    }
    $refreshUri = "$SplunkAPIHost/services/apps/local/$( $SplunkAppName )?refresh=true"
    Invoke-RestMethod -Uri $refreshUri -Method GET -Credential $Script:SplunkCreds -SkipCertificateCheck
}

task Pester SetupVariables, {
    $PesterConfig.Run.Container = (New-PesterContainer -Path "$( $BuildRoot )/test/*.tests.ps1")
    $TestResults = Invoke-Pester -Configuration $Script:PesterConfig
    assert ($TestResults.FailedCount -eq 0) "Pester tests failed."
    assert ($TestResults.FailedContainersCount -eq 0) "Pester container failed."
}

task PesterFrontend SetupVariables, {
    if($TestFrontendInDemo){
        assert(Invoke-WebRequest "http://127.0.0.1:9088") "Demo frontend is not running."
        $Script:SplunkURL = "http://127.0.0.1:9088"
    }
    $TestParams = @{
        SplunkAPIHost = $Script:SplunkAPIHost
        SplunkURL     = $Script:SplunkURL
        DemoTest      = $TestFrontendInDemo
    }
    $FrontendPesterConfig = New-PesterConfiguration
    $FrontendPesterConfig.TestResult.OutputFormat = "NUnitXml"
    $FrontendPesterConfig.TestResult.OutputPath = "$( $BuildRoot )/output/TestResults_Frontend.xml"
    $FrontendPesterConfig.TestResult.Enabled = $True
    $FrontendPesterConfig.Output.Verbosity = "Detailed"
    $FrontendPesterConfig.Run.Container = (New-PesterContainer -Path "$( $BuildRoot )/test/frontend/*.tests.ps1" -Data $TestParams)
    $FrontendPesterConfig.Run.PassThru = $True
    $TestResults = Invoke-Pester -Configuration $FrontendPesterConfig
    assert ($TestResults.FailedCount -eq 0) "Pester tests failed."
    assert ($TestResults.FailedContainersCount -eq 0) "Pester container failed."
}

task EnsureRegularUser SetupVariables, {
    $RequestSplat = @{
        Method               = "GET"
        Uri                  = "$( $SplunkAPIHost )/services/authentication/users/$( $SplunkNormalUser )"
        SkipCertificateCheck = $true
        Credential           = $SplunkCreds
    }
    try {
        $UserExists = Invoke-RestMethod @RequestSplat
    }
    catch {
        $UserExists = $false
    }
    if (-not$UserExists) {
        $RequestSplat = @{
            Method               = "POST"
            Body                 = @{
                name        = $SplunkNormalUser
                password    = $SplunkClearPassword
                roles       = "user"
                output_mode = "json"
            }
            Uri                  = "$( $SplunkAPIHost )/services/authentication/users"
            SkipCertificateCheck = $true
            Credential           = $SplunkCreds
        }
        Invoke-RestMethod @RequestSplat
    }
    else {
        $RequestSplat = @{
            Method               = "POST"
            Body                 = "&roles=user&roles=radware_cwaf_enrichment_admin"
            Uri                  = "$( $SplunkAPIHost )/services/authentication/users/$( $SplunkNormalUser )"
            SkipCertificateCheck = $true
            Credential           = $SplunkCreds
        }
        Invoke-RestMethod @RequestSplat
    }
}

## Frontend Tasks
task BuildFrontend SetupVariables, {
    $frontendBuildPath = "$( $BuildRoot )/frontend"
    Set-Location $frontendBuildPath
    exec {
        yarn install
    }
    exec {
        yarn build
    }
}

task StartDemoApp {
    $demoPath = "$( $BuildRoot )/frontend/packages/radware-enrichment-components"
    Set-Location $demoPath
    exec { yarn start:demo }
}

## Packaging Tasks
task GetVersion {
    $ManifestProps = ((Get-Content -Raw "./splunkapp/app.manifest") | ConvertFrom-JSON)
    $Script:AppVersion = $ManifestProps.info.id.version
    $AppConfVersion = (Select-String -Path "./splunkapp/default/app.conf" -Pattern "^version = ").Line
    $AppConfVersion = $AppConfVersion -replace "version|\s|=", ""

    assert ($AppConfVersion -eq $AppVersion) "Version in app.conf ($AppConfVersion) does not match manifest version ($AppVersion)"

    $Script:PackageFileName = "$( $SplunkAppName )-$( $AppVersion ).tar.gz"
    $Script:AppStageDirectory = "$( $BuildRoot )/output/$( $SplunkAppName )"
}

task BuildSplunkAppTgz GetVersion, {
    #Get-ChildItem -Path "./splunkapp" -Recurse -Filter "__pycache__" | Remove-Item -Recurse -Force
    if (Test-Path -Path $AppStageDirectory) {
        Remove-Item -Path $AppStageDirectory -Recurse -Force
    }
    New-Item -Type Directory -Path $AppStageDirectory
    Copy-Item -Path "$( $BuildRoot )/splunkapp/*" -Destination $AppStageDirectory -Recurse
    $RemoveItems = @(
        "local/"
        "metadata/local.meta"
    )
    foreach ($i in $RemoveItems) {
        $iPath = Join-Path $AppStageDirectory $i
        if (Test-Path $iPath) {
            Remove-Item -Path $iPath -Recurse -Force
        }
    }
    Get-ChildItem -Path $AppStageDirectory -Recurse -Filter "__pycache__" | Remove-Item -Recurse -Force

    Set-Location "$( $BuildRoot )/output/"
    exec { tar --exclude=".*" -czf "$($BuildRoot)/$PackageFileName"  $SplunkAppName }
    Set-Location $BuildRoot
    $PackagePath = (Resolve-Path -Path "$($BuildRoot)/$PackageFileName").Path
    if($env:GITHUB_OUTPUT){
        Write-Output "package_filename=$($PackageFileName)" >> $env:GITHUB_OUTPUT
        Write-Output "package_path=$($PackagePath)" >> $env:GITHUB_OUTPUT
        Write-Host "Running in GitHub - Saving Build Outputs"
    }
    Write-Output "##vso[task.setvariable variable=package_filename]$PackageFileName"
    Write-Output "##vso[task.setvariable variable=package_path]$PackagePath"
}

. (Resolve-Path "$( $BuildRoot )/build/SplunkHelpers.build.ps1")


